cmd/otelinmemexporter/histogram.go

```
@@ -0,0 +1,194 @@
+func explicitBucketsFromHistogramDataPoint(dp pmetric.HistogramDataPoint) []explicitBucket {
+	bucketCounts := dp.BucketCounts().AsRaw()
+	explicitBounds := dp.ExplicitBounds().AsRaw()
+	buckets := make([]explicitBucket, 0, len(bucketCounts))
+	for i, ub := range explicitBounds {
+		buckets = append(buckets, explicitBucket{
+			UpperBound: ub,
+			Count:      bucketCounts[i],
+		})
+	}
+
+	buckets = append(buckets, explicitBucket{
+		UpperBound: math.Inf(+1),
+		Count:      bucketCounts[len(bucketCounts)-1],
+	})
+	return buckets
+}
+
+// deltaExplicitBucketsQuantile calculates the quantile `q` based on the given buckets.
+// The buckets are assumed to be of delta temporality (as opposed to cumulative) and
+// consist of non-overlapping contiguous bounds, e.g. `(-Inf, 0], (0, 10], (10, +Inf]`.
+// The buckets can be unsorted since the function will sort the buckets based on upper
+// bound prior to calculation. The buckets must also have the highest bound of +Inf.
+//
+// The quantile value is interpolated assuming a linear distribution within a bucket.
+// However, if the quantile falls into the highest bucket i.e. `(x, +Inf]`, the lower bound of that
+// bucket (x) is returned . On the other hand, if the quantile falls into the lowest bucket
+// i.e. `(-Inf, y]`, the upper bound of that bucket (y) is returned.
+//
+// Here are some special cases:
+//   - If `q` == NaN, NaN is returned.
+//   - If `q` < 0, -Inf is returned.
+//   - If `q` > 1, +Inf is returned.
+//   - If `buckets` has fewer than 2 elements, NaN is returned.
+//   - If the highest bucket is not +Inf, NaN is returned.
+//   - If `buckets` has 0 observations, NaN is returned.
+func deltaExplicitBucketsQuantile(q float64, buckets []explicitBucket) float64 {
+	if math.IsNaN(q) {
+		return math.NaN()
+	}
+	if q < 0 {
+		return math.Inf(-1)
+	}
+	if q > 1 {
+		return math.Inf(+1)
+	}
+
+	slices.SortFunc(buckets, func(a, b explicitBucket) int {
```

The spec on explicit bounds histogram define that bounds MUST be strictly increasing ([ref](https://github.com/open-telemetry/opentelemetry-proto/blob/a46b66e47a8ca53ae2ac60b9d1b9828c1d5faea4/opentelemetry/proto/metrics/v1/metrics.proto#L477)). So, do we need sorting? Maybe we can fail fast and panic if they are not sorted?

---

That makes sense 👍 Let me update it real quick.